Add @deprecated attribute

132ikl opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->

This PR adds the ability for users to mark commands as deprecated with a `@deprecated` attribute. The attribute can be invoked with or without an extra explanation message. 

Without a message:
![image](https://github.com/user-attachments/assets/359c3f29-dde1-4d21-9501-4ebc9918a156)


With a message:
![image](https://github.com/user-attachments/assets/df3983b1-3ca3-4b60-836b-501d633a3724)


I also changed the deprecated warnings a bit to make them look a bit nicer in my opinion:
- Added an error code (looks nicer with `Warning: ` text)
- Removed command name from error description (it's already in the label)
- Removed suggestion from label and made it optional (it's already in the help text)
- Removed backticks surrounding command names (errors seem to be split on whether they use backticks or not, but we have no special formatting for them which I think leads to unnecessary noise)
- Removed the hardcoded "for more info see" in the help text, so users of the deprecation warning can set the help text however they would like

Here's what a previous deprecation warning looked like, for reference 

![image](https://github.com/user-attachments/assets/39bd6e14-9f7b-4c98-9198-afadb9f8a0cb)

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->
* Adds the `@deprecated` attribute, which can be used to mark a custom command as deprecated. You can optionally supply a message to suggest a new command.

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->
Changed the attributes completions test to not be hard-coded (so if we add any more attributes they don't need to be manually added) and added a test for `@deprecated`.

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->
N/A

Bahex commented:
This only shows the warnings when deprecated commands' definitions are parsed and not on _calls_ to the deprecated commands right?

I like the idea of `@deprecated` and the refactors you made, but I think we first need a way to properly handle deprecation warnings. If we decide to deprecate a command in std and use `@deprecated` as it is now, rather than users getting warned when they use deprecated commands, they will be warned when they import modules that contain deprecated commands.

I looked around a bit to see what's usually done when built-in commands are deprecated and they just throw a warning _each time_ they are called and then run as usual.
I think we need a different approach where deprecated commands (let's say they are just marked by changing their category to deprecated) throw a warning when they are _first_ called. And by _called_ I don't mean when they run, I mean when they appear in parsed code.

```nushell
# aaa.nu
@deprecated
export def foo [] { ... }
```
```nushell
# bbb.nu
use aaa.nu
export def bar [] {
    foo
}
``` 
```nushell
# repl

❯ use aaa.nu
# no warning

❯ use bbb.nu
# warning is thrown since
```

132ikl commented:
> they will be warned when they import modules that contain deprecated commands

whoops, not sure how that slipped my mind. this definitely needs to happen when parsing a call, not when parsing the definition

> throw a warning when they are first called

this is definitely needed, we discussed this before in the context of `table`->`render` and also lints, although i don't think we need that for an initial implementation necessary

> let's say they are just marked by changing their category to deprecated

i think we need something more complex than this so we can have a message still. maybe `@deprecated` can just be a normal attribute, and then we check for the `@deprecated` attribute when we parse a call?

Bahex commented:
Yeah, a normal attribute makes more sense than a category for this since it can also store a message.

We can add a `Command::is_deprecated()` that for `BlockCommand`'s is based on their attributes and can be manually implemented for built-ins.

132ikl commented:
just updated the PR with your suggest @Bahex. also added a "report log" which lets us make sure we only report deprecation warnings (and later, lints if we decide to add those) once. It needs to be in the reporting code since we might generate the warning without reporting it (for example, while typing the command in the REPL). I'm not sure about storing the `DeclId` in the error itself, but it seems to be the only place which really makes sense

Bahex reviewed (approved):
Wow I didn't expect this to be done so soon :rocket:

Everything looks goods, and it works just as expected:
- reporting just once in shell
- reporting each occurrence in lsp

![image](https://github.com/user-attachments/assets/3daef445-e159-45de-a7d5-584059809f8b)

cptpiepmatz commented on crates/nu-protocol/src/deprecation.rs:
```diff
@@ -0,0 +1,28 @@
+use crate::{ParseWarning, ast::Call};
+
+#[derive(Clone)]
+pub enum DeprecationStatus {
+    Undeprecated,
+    Deprecated(Option<String>),
+}
```
I don't like this enum as it is. `Undeprecated` sounds weird to me, I think we could better work with an `Option` instead. Also only having `Deprecated` is not granular enough. We talked about different deprecation levels, this doesn't provide that.

I'm also thinking about stronger deprecation entries, they could include:
- deprecation level, how often to notify the user
- since when it is deprecated, a version maybe
- expected removal if even
- reason for it
- help not

Only having a string incentivizes to weak messaging.

132ikl replied:
I added all of these except "reason for it", since I figured that would go in the help string and it's not really clear where else that would fit into the `miette` error

cptpiepmatz commented on crates/nu-protocol/src/errors/cli_error.rs:
```diff
@@ -11,6 +16,9 @@ use miette::{
 use thiserror::Error;
 
+/// The thread-global log of reported errors and warnings, so specific reports are only shown once
+static REPORT_LOG: LazyLock<Mutex<ReportLog>> = LazyLock::new(Mutex::default);
```
I know I suggested the `LazyLock` but now I'm unsure if we should have it in the `EngineState` instead. We have access to it basically everywhere and this static doesn't work well, when you run with multiple engines, which is not relevant to us but maybe embedders.

132ikl replied:
will be doing this as discussed in meeting

cptpiepmatz commented on crates/nu-protocol/src/errors/cli_error.rs:
```diff
@@ -20,6 +28,31 @@ struct CliError<'src>(
+    } {
+        let is_first_report = REPORT_LOG
+            .lock()
+            .expect("report log is poisioned")
```
In our dependecy graph is already `parking_lot`, it provides a Mutex which doesn't poison. Maybe that's better.

132ikl replied:
no longer relevant after moving to engine state

cptpiepmatz commented on crates/nu-protocol/src/engine/command.rs:
```diff
@@ -133,6 +134,10 @@ pub trait Command: Send + Sync + CommandClone {
         self.command_type() == CommandType::Plugin
     }
 
+    fn deprecation_status(&self) -> DeprecationStatus {
+        DeprecationStatus::Undeprecated
+    }
+
```
While this is good for the whole command, we also have deprecations for flags and stuff, how do we deal with that?

Bahex replied:
That's a really good point, which I somehow forgot despite deprecating `get --sensitive` recently.

132ikl replied:
added a `DeprecationType` enum which lets us mark flags as deprecated. there's also the idea that we might need deprecations we can only detect at run-time which we might be able to work into the `DeprecationType` enum, and then make the check in `parse_warning` return `false` for all of those. it's not exactly clear to me when we would actually trigger these deprecations, although I guess we would need a more specific example first? anyways, we can add that later if we need it

132ikl commented:
this is almost ready, just need to move the `ReportLog` to `EngineState` and also investigate how difficult it would be to expose the deprecation stuff to plugins

cptpiepmatz commented on crates/nu-derive-value/src/from.rs:
```diff
@@ -590,6 +589,12 @@ fn parse_value_via_fields(
+                    (false, true) => quote! {
+                        #ident: record
+                            .remove(#ident_s)
+                            .and_then(|v| <#ty as nu_protocol::FromValue>::from_value(v).ok())
```
I don't like that this impl would would silently fallback to the default. If that field exists but cannot be converted, it should throw an error.

cptpiepmatz commented on crates/nu-cmd-base/src/wrap_call.rs:
```diff
@@ -0,0 +1,101 @@
+///
+/// A new `WrapCall` instance has to be returned after each function to ensure
+/// that there is only ever one copy of mutable [`Stack`] reference.
+pub enum WrapCall<'a> {
```
That type feels odd to me, I think we can come up with a better way of implementing this. But I don't see this as a blocker rn.

cptpiepmatz commented on crates/nu-protocol/src/deprecation.rs:
```diff
@@ -0,0 +1,142 @@
+    #[nu_value(rename = "flag", default)]
+    ty: DeprecationType,
+    #[nu_value(rename = "report")]
+    pub report_mode: ReportMode,
```
Why is only this field `pub`?

132ikl replied:
no idea, I just made the rest of them pub since it's a pub struct (which I think is probably reasonable?)

c2136a4b9

132ikl commented on crates/nu-protocol/src/deprecation.rs:
```diff
@@ -0,0 +1,142 @@
+impl FromValue for ReportMode {
+    fn from_value(v: Value) -> Result<Self, ShellError> {
+        let span = v.span();
+        let Value::String { val, .. } = v else {
+            return Err(ShellError::CantConvert {
+                to_type: Self::expected_type().to_string(),
+                from_type: v.get_type().to_string(),
+                span: v.span(),
+                help: None,
+            });
+        };
```
...not sure what i was cooking here

132ikl commented on crates/nu-protocol/src/deprecation.rs:
```diff
@@ -0,0 +1,142 @@
+use crate::{self as nu_protocol, ReportMode, Span};
+
+#[derive(FromValue)]
+pub struct DeprecationEntry {
```
good point, added in 49036f527

132ikl commented on crates/nu-protocol/src/errors/parse_warning.rs:
```diff
@@ -1,27 +1,64 @@
+        label: String,
+        report_mode: ReportMode,
+        help: String,
     },
```
aha, I thought there was a way to do this but I forgot! fixed in a82548e55

132ikl commented on crates/nu-protocol/src/deprecation.rs:
```diff
@@ -0,0 +1,142 @@
+// The `FromValue` derive macro fully qualifies paths to "nu_protocol".
+use crate::{self as nu_protocol, ReportMode, Span};
+
+#[derive(FromValue)]
+pub struct DeprecationEntry {
```
yes, it's how `@deprecated` records are converted into `DeprecationEntry`, [used here](https://github.com/nushell/nushell/tree/a82548e55731cf40966f88c5129ad9f18b6a5461/crates/nu-protocol/src/signature.rs#L796)

cptpiepmatz reviewed (approved):
Great work, thank you ♥️
